azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+        # dont have this logic in place yet.
+        assert stage.next.run_op.call_count == 3
+        assert stage.next.run_op.call_args_list[1][0][0] == first_fake_op
+        assert stage.next.run_op.call_args_list[2][0][0] == second_connect
+
+        # now, complete second_connect to give second_fake_op a chance to get passed down
+        operation_flow.complete_op(stage=stage.next, op=second_connect)
+        assert stage.next.run_op.call_count == 4
+        assert stage.next.run_op.call_args_list[3][0][0] == second_fake_op
+
+    @pytest.mark.parametrize(
+        "params",
+        [
+            pytest.param(
+                {
+                    "pre_connected_flag": True,
+                    "first_connect_op": pipeline_ops_base.DisconnectOperation,
+                    "mid_connect_flag": False,
+                    "second_connect_op": pipeline_ops_base.DisconnectOperation,
+                },
+                id="Disconnect followed by Disconnect",
+            ),
+            pytest.param(
+                {
+                    "pre_connected_flag": False,
+                    "first_connect_op": pipeline_ops_base.ConnectOperation,
+                    "mid_connect_flag": True,
+                    "second_connect_op": pipeline_ops_base.ConnectOperation,
+                },
+                id="Connect followed by Connect",
+            ),
+            pytest.param(
+                {
+                    "pre_connected_flag": True,
+                    "first_connect_op": pipeline_ops_base.ReconnectOperation,
+                    "mid_connect_flag": True,
+                    "second_connect_op": pipeline_ops_base.ConnectOperation,
+                },
+                id="Reconnect followed by Connect",
+            ),
+        ],
+    )
+    @pytest.mark.it(
+        "Immediately completes a second op which was waiting for a first op that succeeded"
+    )
+    def test_immediately_completes_second_op(self, stage, params, mocker):
+        first_connect_op = params["first_connect_op"](mocker.MagicMock())
+        second_connect_op = params["second_connect_op"](mocker.MagicMock())
+        stage.pipeline_root.connected = params["pre_connected_flag"]
+
+        stage.run_op(first_connect_op)
+        stage.run_op(second_connect_op)
+
+        # first_connect_op has been passed down.  second_connect_op is waiting for first disconnect to complete.
+        assert stage.next.run_op.call_count == 1
+        assert stage.next.run_op.call_args[0][0] == first_connect_op
+
+        # complete first_connect_op
+        stage.pipeline_root.connected = params["mid_connect_flag"]
+        operation_flow.complete_op(stage=stage.next, op=first_connect_op)
+
+        # second connect_op should be completed without having been passed down.
+        assert stage.next.run_op.call_count == 1
+        assert_callback_succeeded(op=second_connect_op)
```

Comment:
I'd also like to see a test case for the case we discussed offline the other day:

1) Connection operations triggers op serialization and starts block
2) Several operations pile up in the queue
3) Connection operation completes, queue is unblocked
4) While in the process of emptying the queue, a new operation is sent down

As per our discussion, you said this new operation should be blocked, or not allowed to be sent down, or something to that effect, until the queue has successfully emptied.

Reply:
Added in TFS task 5117988

---
In reply to: [312679508](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312679508) [](ancestors = 312679508)